domain/src/main/java/org/oppia/android/domain/snackbar/SnackbarController.kt

```diff
@@ -0,0 +1,86 @@
+package org.oppia.android.domain.snackbar
+
+import androidx.annotation.StringRes
+import com.google.common.util.concurrent.SettableFuture
+import kotlinx.coroutines.Deferred
+import org.oppia.android.util.data.AsyncDataSubscriptionManager
+import org.oppia.android.util.data.DataProvider
+import org.oppia.android.util.data.DataProviders
+import java.util.LinkedList
+import java.util.Queue
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val GET_CURRENT_SNACKBAR_REQUEST_PROVIDER_ID =
+  "get_current_snackbar_request_provider_id"
+
+/** Controller for enqueueing, dismissing, and retrieving snackbars. */
+@Singleton
+class SnackbarController @Inject constructor(
+  private val dataProviders: DataProviders,
+  private val asyncDataSubscriptionManager: AsyncDataSubscriptionManager,
+) {
+
+  val dismissFuture = SettableFuture.create<Unit>()
+  private val _snackbarRequestQueue: Queue<ShowSnackbarRequest> = LinkedList()
+
+  /** Queue of the snackbar requests that are to be shown based on FIFO. */
+  val snackbarRequestQueue: Queue<ShowSnackbarRequest>
+    get() = _snackbarRequestQueue
+
+  val currentState = CurrentSnackbarState.NotShowing
+
+  fun getCurrentSnackbarState(): DataProvider<CurrentSnackbarState> {
+
+    val currentRequest = _snackbarRequestQueue.peek()
+
+    if (_snackbarRequestQueue.isEmpty()) {
+      return dataProviders.createInMemoryDataProvider(CurrentSnackbarState.NotShowing) {
+        return@createInMemoryDataProvider CurrentSnackbarState.NotShowing
+      }
+    }
+  }
+
+  fun enqueueSnackbar(request: ShowSnackbarRequest) {
+    _snackbarRequestQueue.add(request)
+    notifyPotentialSnackbarChange()
+  }
+
+  fun notifySnackbarShowing(snackbarId: Int, onShow: Deferred<Unit>, onDismiss: Deferred<Unit>) {
+    // onDismiss is resolved when the snackbar by unique ID snackbarId is no longer showing.
+    CurrentSnackbarState.NotShowing
+
+//    val showFuture = onShow.await()
```

Comment:
Here we need for onShow to finish before we can actually change the snackbar to be shown. We also need to observe onDismiss so that we know when to remove the snackbar from the queue. Something like:

```kotlin
CoroutineScope(backgroundDispatcher).launch {
  onShow.await()
  // TODO: Move to helper.
  val queueFrontWhenShown = snackbarQueue.peek()
  check(queueFrontWhenShown is CurrentSnackbarState.WaitingToShow)
  check(queueFrontWhenShown.snackbarId == snackbarId)
  snackbarQueue.removeFirst()
  snackbarQueue.pushFirst(CurrentSnackbarState.Showing(queueFrontWhenShown.request, snackbarId))
  asyncDataSubscriptionManager.notifyChange(GET_CURRENT_SNACKBAR_REQUEST_PROVIDER_ID)
  onDismiss.await()
  // TODO: Move to helper.
  val queueFrontWhenDismissed = snackbarQueue.peek()
  check(queueFrontWhenDismissed is CurrentSnackbarState.Showing)
  check(queueFrontWhenDismissed.snackbarId == snackbarId)
  snackbarQueue.removeFirst()
  asyncDataSubscriptionManager.notifyChange(GET_CURRENT_SNACKBAR_REQUEST_PROVIDER_ID)
}.invokeOnCompletion {
  // Log error if it != null.
}
```

Reply:
@BenHenning when on this line   `snackbarQueue.pushFirst(CurrentSnackbarState.Showing(queueFrontWhenShown.request, snackbarId))`
actually, this is not compatible as snackbarRequestQueue is of type `ShowSnackbarRequest` and we can't add `CurrentSnackbarState`, can you PTAL?

![image](https://github.com/oppia/oppia-android/assets/92685493/4b5db2a9-4732-446b-a0bb-1343aa2975a0)